helm-chart/binderhub/values.yaml

```diff
@@ -63,83 +63,14 @@ jupyterhub:
         auth_enabled: false
     extraConfig:
       00-binder: |
-        from tornado import web
+        # This must be embedded since the Z2JH hub container won't contain BinderHub
+        from binderhub.binderspawner_mixin import BinderSpawnerMixin
```

Comment:
Would it make sense to do this embedding during the helm packaging step where the spawner file exist in `helm-chart/binderhub/files/` and is injected using the [`extraFiles` directive of z2jh](https://zero-to-jupyterhub.readthedocs.io/en/latest/resources/reference.html?highlight=extraFiles#hub-extrafiles) ?

Reply:
I think that requires passing `--set-file` on the Helm command line, as far as I know it can't be set in `values.yaml`.

Helm does support [embedding the content of files using `Files.Get`](https://helm.sh/docs/chart_template_guide/accessing_files/) which would be ideal and could be added to the Z2JH chart, but unfortunately it only works for files within the same chart- a file can't be passed from the parent to the child chart :disappointed: https://github.com/helm/helm/issues/1892

Reply:
You can make a configmap part of the binderhub helm chart, then use extraVolumes/extraVolumeMounts though - right? Anything you put in a specific folder will be loaded as well automatically by jupyterhub as config.

Ita documented under hub.extraFiles what that folder was. I recall it was /usr/local/jupyterhub_config.d/ or similar.

Reply:
Doesn't a configmap still require embedding the py file as yaml?

Reply:
No, files.get!

Reply:
BTW this is pretty much what we do with the [PersistentBinderSpawner class](https://github.com/gesiscss/persistent_binderhub/blob/master/persistent_binderhub/files/jupyterhub/persistent_bhub_config.py) (embed with  [extraVolumes/extraVolumeMounts](https://github.com/gesiscss/persistent_binderhub/blob/6b1b7995241a9f0cdb7a74713a889e83beffe57a/persistent_binderhub/values.yaml#L89-L93)) too.

Reply:
Also I was looking into this sometime back, and I was able to make the tests happy with https://github.com/jupyterhub/binderhub/compare/master...MridulS:persistent_bhub (the final aim of the branch is to send in the PR which augments the BinderSpawner to work with Persistence too 😄)

Reply:
Im on mobile, but z2jh has a standalone .py file it loads into a configmap (as a string value but not hardcoded).

Reply:
I've just tried this, but it means the JupyterHub and BinderHub charts have to be deployed simultaneously since the configmap is part of the BinderHub chart, but mounted by the JupyterHub chart.

Our CI tests assume Z2JH can be deployed independently by extracting the `jupyterhub` chart config from the `binderhub/values.yaml`
- https://github.com/jupyterhub/binderhub/blob/8c8c0be2aba6542849269f1de440697a614ce9f5/.github/workflows/test.yml#L87-L90
- https://github.com/jupyterhub/binderhub/blob/8c8c0be2aba6542849269f1de440697a614ce9f5/testing/local-binder-k8s-hub/install-jupyterhub-chart#L35-L37

We could manually create the required configmap? The alternative which I'm investigating is to keep everything in `values.yaml` and have `setup.py` pull it out instead.

Reply:
> Our CI tests assume Z2JH can be deployed independently by extracting the jupyterhub chart config from the binderhub/values.yaml

Oh, ah... Okay.

Well, you could do `helm template ... --show-only templates/configmap.yaml | kubectl apply -f -` as an ugly hack then, I don't see another great solution if we need to deploy binderhub logic related to jupyterhub in k8s but not the entire binderhub in a k8s context.

Reply:
The overall problem is it's JupyterHub specific but not k8s specific.

What do you think of the latest changes? I've put the config back into `values.yaml`, and I'm copying it into the binderhub python source in setup.py

It introduces a dependency on `pyyaml` to run `setup.py`, but given that `setup.py` already has external requirements like `npm` I don't think that matters too much.

Reply:
I understand, I think this can be marked as resolved given that we add notes about this complexity. I've made code suggestions below about such notes - feel free to tweak them!